virtual_rainforest/models/soil/env_factors.py

```diff
@@ -0,0 +1,306 @@
+
+    This factor impacts enzyme saturation constants, based on the assumption that finely
+    textured soils will restrict enzyme access to available C substrates (which protects
+    them). Its form is taken from :cite:t:`fatichi_mechanistic_2019`.
+
+    Args:
+        clay_fraction: The fraction of the soil which is clay [unitless]
+        base_protection: The protection that a soil with no clay provides [unitless]
+        protection_with_clay: The rate at which protection increases with increasing
+           clay [unitless]
+
+    Returns:
+        A multiplicative factor capturing how much the protection due to soil structure
+        changes the effective saturation constant by [unitless]
+    """
+
+    return base_protection + protection_with_clay * clay_fraction
+
+
+def calculate_clay_impact_on_necromass_decay(
+    clay_fraction: NDArray[np.float32], decay_exponent: float
+) -> NDArray[np.float32]:
+    """Calculate the impact that soil clay has on necromass decay to LMWC.
+
+    Necromass which doesn't breakdown fully gets added to the POM pool instead.
+
+    Args:
+        clay_fraction: The fraction of the soil which is clay [unitless]
+        sorption_exponent: Controls the impact that differences in soil clay content
+            have on the proportion of necromass that decays to LMWC [unitless]
+
+    Returns:
+        A multiplicative factor capturing the impact that soil clay has on the
+        proportion of necromass decay which sorbs to form POM [unitless]
+    """
+
+    return np.exp(decay_exponent * clay_fraction)
+
+
+def calculate_leaching_rate(
+    solute_density: NDArray[np.float32],
+    vertical_flow_rate: NDArray[np.float32],
+    soil_moisture: NDArray[np.float32],
+    solubility_coefficient: float,
+    soil_layer_thickness: float,
+) -> NDArray[np.float32]:
+    """Calculate leaching rate for a given solute based on flow rate.
+
+    This functional form is adapted from :cite:t:`porporato_hydrologic_2003`. The amount
+    of solute that is expected to be found in dissolved form is calculated by
+    multiplying the solute density by its solubility coefficient. This is then
+    multiplied by the frequency with which the water column is completely replaced, i.e.
+    the ratio of vertical flow rate to water column height.
+
+    Args:
+        solute_density: The density of the solute in the soil [kg solute m^-3]
+        vertical_flow_rate: Rate of flow downwards through the soil [mm day^-1]
+        soil_moisture: Volumetric relative water content of the soil [unitless]
+        solubility_coefficient: The solubility coefficient of the solute in question
+            [unitless]
+        soil_layer_thickness: Thickness of the biogeochemically active soil layer [m]
+
+    Returns:
+        The rate at which the solute in question is leached [kg solute m^-3 day^-1]
+    """
+
+    # Vertical flow rate has to be converted into m day^-1
+    vert_flow_meters = vertical_flow_rate / 1e3
+
```

Comment:
We should have a chat about this, especially the units. I included a lot of conversions to go from m/s in a soil column to mm per month per area, which might all be unnecessary sources of error if you don't need that.